code/modules/projectiles/projectile.dm

```diff
@@ -751,191 +764,358 @@
-		after_z_change(old, loc)
-		if(!hitscanning)
-			pixel_x = trajectory.return_px()
-			pixel_y = trajectory.return_py()
-		forcemoved = TRUE
-		hitscan_last = loc
-		SEND_SIGNAL(src, COMSIG_PROJECTILE_PIXEL_STEP)
-	if(QDELETED(src)) //deleted on last move
-		return
-	if(!hitscanning && !forcemoved)
-		pixel_x = trajectory.return_px() - trajectory.mpx * trajectory_multiplier * SSprojectiles.global_iterations_per_move
-		pixel_y = trajectory.return_py() - trajectory.mpy * trajectory_multiplier * SSprojectiles.global_iterations_per_move
-		animate(src, pixel_x = trajectory.return_px(), pixel_y = trajectory.return_py(), time = 1, flags = ANIMATION_END_NOW)
-	Range()
 
-/obj/projectile/proc/process_homing() //may need speeding up in the future performance wise.
+		var/distance_to_move = min(distance_to_border, pixels_to_move)
+		// For homing we cap the maximum distance to move every loop
+		if (homing && distance_to_move > SSprojectiles.pixels_per_decisecond)
+			distance_to_move = SSprojectiles.pixels_per_decisecond
+
+		// Figure out if we move to the next turf and if so, what its positioning relatively to us is
+		var/x_shift = SIGN(movement_vector.pixel_x) * (!isnull(x_to_border) && distance_to_move >= x_to_border)
+		var/y_shift = SIGN(movement_vector.pixel_y) * (!isnull(y_to_border) && distance_to_move >= y_to_border)
+		var/turf/new_turf = locate(x + x_shift, y + y_shift, z)
+		// Calculate where in the turf we will be when we cross the edge, for impact VFX
+		entry_x = pixel_x + distance_to_move * movement_vector.pixel_x - ICON_SIZE_X * x_shift
+		entry_y = pixel_y + distance_to_move * movement_vector.pixel_y - ICON_SIZE_Y * y_shift
+		var/delete_distance
+
+		if (x_shift || y_shift)
+			// We've hit an invalid turf, end of a z level or smth went wrong
+			if (!istype(new_turf))
+				qdel(src)
+				return
+
+			// Move to the next tile
+			step_towards(src, new_turf)
+			SEND_SIGNAL(src, COMSIG_PROJECTILE_MOVE_PROCESS_STEP)
+			// We hit something and got deleted, stop the loop
+			if (QDELETED(src))
+				return
+			// If we've impacted something, we need to animate our movement until the actual hit
+			// Otherwise the projectile visually disappears slightly before the actual impact
+			if (deletion_queued)
+				delete_distance = clamp(distance_to_move + sqrt(abs(entry_x - impact_x) ** 2 + abs(entry_y - impact_y) ** 2), 0, pixels_to_move)
+
+		// We cannot move more than one turf worth of distance per loop, so this is a safe solution
+		pixels_moved_last_tile += distance_to_move
+		if (!deletion_queued && pixels_moved_last_tile >= ICON_SIZE_ALL)
+			reduce_range()
+			if (QDELETED(src))
+				return
+			// Similarly with range out deletion, need to calculate how many pixels we can actually move before deleting
+			if (deletion_queued)
+				delete_distance = distance_to_move - (pixels_moved_last_tile - ICON_SIZE_ALL)
+
+		if (deletion_queued)
+			//pixel_x -= x_shift * ICON_SIZE_X
+			//pixel_y -= y_shift * ICON_SIZE_Y
```

Comment:
?

Reply:
was needed before i swapped it to end_now, forgor